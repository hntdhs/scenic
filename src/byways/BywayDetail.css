.comment-user {
    color: green;
}

.comment-text {
    font-size: medium;
}

.comment-area {
    max-width: 50%;
    border: black;
}