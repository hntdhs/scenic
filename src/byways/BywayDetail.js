import React, { useState, useEffect, useContext } from "react";
import { useParams } from "react-router-dom";
import BywayApi from "../api/api";
import CommentForm from "../actions/CommentForm";
import FavoriteAByway from "../actions/FavoriteAByway";
import UserContext from "../auth/UserContext";
import moment from 'moment';
// see Moment docs on display for how to format comment dates/times in different ways
import { useToasts } from 'react-toast-notifications';


function BywayDetail() {
    const { name } = useParams();
    const { currentUser } = useContext(UserContext);
    const { addToast } = useToasts();
    const [byway, setByway] = useState(null);
    const [comments, setComments] = useState([]);
    const [isFavorite, setIsFavorite] = useState(false);


    useEffect(() => {        
        async function getByway(name) {
            try {
                const b = await BywayApi.getByway(name)
                setByway(b);
                const c = await BywayApi.getCommentsByByway(b.id);
                setComments(c);
                const favorites = await BywayApi.getUserFavorites(currentUser.username, "name", "asc");

                if (favorites.find(i => i.byway_id === b.id)) {
                    setIsFavorite(true);
                } 
        } catch (errors) {
            if (errors.length > 0) {
                addToast(errors[0], { appearance: 'error' });
            }
            }
        }    

        getByway(name);
    }, [name]);


    // empty array makes it only happen first time

    const addComment = (comment) => {
        const updatedComments = [...comments, comment]
        setComments(updatedComments);
    }

    function checkGeoFeaturesLength(byway) {
        let tempString = byway.geographicFeatures
        tempString = tempString[0].toUpperCase() + tempString.slice(1)
        const splitted = tempString.split(',')
        if (splitted.length >= 3) {
            // if byway has 3 or more - features = Feature, feature and feature
            // return byway.geographicFeatures.charAt(0).toUpperCase() + byway.geographicFeatures.replace(/("[^"]+"|\w+)$/, "and $1").slice(1);
            const comma_separated_list_until_last_element = splitted.slice(0, splitted.length - 1).join(', ')
            // takes the splitted string, slices it from beginning up until the last element and joins them on comma, but there won't be a comma after the last feature, meaning we can add the last feature from splitted and return it at the end of this below with 'and' in front of it but no comma
            return comma_separated_list_until_last_element + ' and ' + splitted[splitted.length - 1]
            //return byway.geographicFeatures.charAt(0).toUpperCase() + byway.geographicFeatures.slice(1, lastIndex) + ' and' + byway.geographicFeatures(lastIndex + 1);
        } else if (splitted.length === 2) {
            // if byway has 2 - features = Feature and feature
            // this replaced the comma with regex - return byway.geographicFeatures.charAt(0).toUpperCase() + byway.geographicFeatures.replace(/("[^"]+"|\w+)$/, "and $1").slice(1);
            const newString = splitted[0] + ' and' + splitted[1]
            return newString.charAt(0).toUpperCase() + newString.slice(1)
        } else if (splitted.length === 1) {
            // if byway has 1 - features = Feature
            // byway with one feature (charles street) showed 'no features verified'
            // after adding the .length to byway.geoFeatures charles street shows '3 or more'
            return splitted[0]
        } else {
            // if byway has none - 'No geographic features have been verified for this byway'
            // byway with no features (hallowed ground) showed 'no features verified', maybe just because things are falling down to this part of the function
            // after adding the .length to byway.geoFeatures this shows 'no features verified'
            return "No geographic features have been verified for this byway.";

        }
        
    }

    return (
        <div>
            {byway
                ? (
                    <div>
                        <h1>{byway.name}</h1>
                        <img src={byway.image} alt={byway.name}></img>
                        <h2>{byway.state}</h2>
                        <h3>{byway.length} miles / { Math.round((parseFloat(byway.length) * 1.60934)) } kilometres</h3>
                        <p><b>Fees:</b> {byway.fees}</p>
                        {<p><b>Wondering what kind of natural settings you'll see on this byway?</b> {checkGeoFeaturesLength(byway)}.</p> }
                        <p><b>About {byway.name}</b><br></br>{byway.description}</p>
                        {isFavorite === true
                        ? (
                            <h4>Byway is in your favorites</h4>
                        ) : (
                            <FavoriteAByway id={byway.id} />
                        )
                        }
                        <h4>Comment on {byway.name}</h4>
                        <CommentForm name={byway.name} id={byway.id} onAdd={addComment}/>
                        <div>
                            {comments
                            ? (
                                <div>
                                    {comments.map((c, i) => (
                                        <div key={i}>
                                            <h4>User <a href={`/profile/${c.username}`}>{c.username}</a> said this at {moment(c.create_at).format("MM/DD/YYYY hh:mm a")}:</h4>
                                            <p>{c.comment}</p>
                                        </div>
                                    ))}
                                </div>
                            ) : (
                                <p>no one has commented</p>
                            )
                            }
                        </div>

                    </div>
                    
                ) : (
                    <h1>no byway by that name</h1>
                )
            }
        </div>
    )
}

export default BywayDetail;