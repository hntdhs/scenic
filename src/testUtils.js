import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  photo_url: null,
};
// const
let UserProvider =
    ({ children, currentUser = demoUser }) => (
    <UserContext.Provider value={{ currentUser }}>
      {children}
    </UserContext.Provider>
    )


// const UserProvider =
//     ({ children, currentUser = demoUser, hasAppliedToJob = () => false }) => (
//     <UserContext.Provider value={{ currentUser, hasAppliedToJob }}>
//       {children}
//     </UserContext.Provider>
// );

export { UserProvider };
