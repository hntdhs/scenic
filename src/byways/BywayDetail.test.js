import React from "react";
import { render } from "@testing-library/react";
import BywayDetail from "./BywayDetail";
// when importing needed components from other files, if they're 'export default they can be named anything here. so this could be import Byway or import BywayDetail or anything else
import { MemoryRouter, Route } from "react-router-dom";
import { UserProvider } from "../testUtils";
import { ToastProvider } from 'react-toast-notifications';


it("renders without crashing", function () {
    render(
      <ToastProvider>
        <MemoryRouter>
          <UserProvider>
            <BywayDetail />
          </UserProvider>
        </MemoryRouter>
      </ToastProvider>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <ToastProvider>
        <MemoryRouter initialEntries={["/byways/Merritt%20Parkway"]}>
          {/* starting from the route given in initialEntries, MemoryRouter is necessary, see line 17 of comp detail - const { handle } = useParams(); - it needs a handle because the component is taking the company from the parameters */}
          {/* just storing the rendering in memory since nothing is actually rendering it */}
          <UserProvider>
            {/* feeding in current user info */}
            <Route path="/byways/:name">
              <BywayDetail name="Merritt%20Parkway" state="Connecticut" length="37" geographic_features="Forest"/>
            </Route>
          </UserProvider>
        </MemoryRouter></ToastProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <ToastProvider>
        <MemoryRouter initialEntries={["/byways/Arroyo%20Seco%20Historic%20Parkway"]}>
          <UserProvider>
            <Route path="/byways/:name">
              <BywayDetail name="Arroyo%20Seco%20Historic%20Parkway" state="California" length="9" designation="National Scenic Byway" fees="none" image="https://fhwaapps.fhwa.dot.gov/bywaysp/uploads/asset_files/2176/36428_S20082-04_sq.jpg" description="The Arroyo Seco Parkway connects Los Angeles and Pasadena through the historic Arts and Crafts landscape of the Arroyo Seco. Conceived in the parkway tradition with gentle curves, lush landscaping, and scenic vistas, the byway incorporated the modern elements that would lay the groundwork for the California freeway system." geographic_features=""/>
            </Route>
          </UserProvider>
        </MemoryRouter></ToastProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <ToastProvider>
        <MemoryRouter initialEntries={["/byways/Talladega%20Scenic%20Drive"]}>
          <UserProvider>
            <Route path="/byways/:name">
              <BywayDetail name="Talladega%20Scenic%20Drive" state="Alabama" length="26.4" designation="National Scenic Byway" fees="There is a $1.00 per person fee to take an optional side trip into the Cheaha State Park." image="https://fhwaapps.fhwa.dot.gov/bywaysp/uploads/asset_files/2057/67321_talladega_primary_photo_sq.jpg" description="Talladega Drive offers a bird's-eye view of scenic mountains, rock outcroppings, and small rural settlements within the Talladega National Forest. Heading towards Cheaha State Park and Alabama's highest peak, with an elevation of 2,407 feet, travel the backbone of Horseblock and Cheaha Mountains, the southernmost extension of the Appalachian Mountains." geographic_features="Forest and mountains"/>
            </Route>
          </UserProvider>
        </MemoryRouter></ToastProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <ToastProvider><MemoryRouter initialEntries={["/byways/Talladega%20Scenic%20Drive"]}>
          <UserProvider>
            <Route path="/byways/:name">
              <BywayDetail name="Glenn%20Highway" state="Alaska" length="135" designation="National Scenic Byway" fees="None" image="https://fhwaapps.fhwa.dot.gov/bywaysp/uploads/asset_files/2483/36398_S20074-16_sq.jpg'" description="Tracing the receding glaciers responsible for this rugged gateway to Alaska's interior, Glenn Highway tells of powerful geological processes and the resourceful people who have managed to thrive along its corridor. Begin your journey in Anchorage and wind along 135 miles through fascinating landscapes, historical sites, and cozy roadhouses." geographic_features="Mountains, forest, and river"/>
            </Route>
          </UserProvider>
        </MemoryRouter></ToastProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
