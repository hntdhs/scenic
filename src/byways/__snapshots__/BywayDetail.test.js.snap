// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      no byway by that name
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 2`] = `
<DocumentFragment>
  <div>
    <h1>
      no byway by that name
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 3`] = `
<DocumentFragment>
  <div>
    <h1>
      no byway by that name
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 4`] = `
<DocumentFragment>
  <div>
    <h1>
      no byway by that name
    </h1>
  </div>
</DocumentFragment>
`;
